app/models.py

```
@@ -49,6 +49,153 @@ class TurnConfig(BaseModel):
+        le=50_000,
+        description="Optional per-source token cap before the request-level cap is applied",
+    )
+
+    @field_validator("url")
+    @classmethod
+    def validate_url(cls, value: Optional[str]) -> Optional[str]:
+        if value is None:
+            return value
+        normalized = value.strip()
+        if not normalized.startswith(("http://", "https://")):
+            raise ValueError("prompt data source url must start with http:// or https://")
```

Ruff (0.15.20)

[warning] Avoid specifying long messages outside the exception class

(TRY003)